File: golang/codec/message_data.go
package codec

import (
	"encoding/binary"
	"fmt"
)

type DataMessage struct {
	ChannelID uint32
	Length    uint32
	Data      []byte
}

func (msg DataMessage) String() string {
	return fmt.Sprintf("{DataMessage ChannelID:%d Length:%d Data: ... }",
		msg.ChannelID, msg.Length)
}

func (msg DataMessage) Channel() (uint32, bool) {
	return msg.ChannelID, true
}

func (msg DataMessage) MarshalMux() ([]byte, error) {
	packet := make([]byte, payloadSizes[msgChannelData]+1)
	packet[0] = msgChannelData
	binary.BigEndian.PutUint32(packet[1:5], msg.ChannelID)
	binary.BigEndian.PutUint32(packet[5:9], msg.Length)
	return append(packet, msg.Data...), nil
}

func (msg *DataMessage) UnmarshalMux(b []byte) error {
	msg.ChannelID = binary.BigEndian.Uint32(b[1:5])
	msg.Length = binary.BigEndian.Uint32(b[5:9])
	msg.Data = b[9:]
	return nil
}

File: csharp/qmux/codec/message_data.cs
namespace qmux.codec;

using System.IO;
using System.Text;

public struct DataMessage : IMessage
{
    private nuint channelId;
    private nuint length;
    private byte[] data;

    public string String()
    {
        return $"DataMessage ChannelId:{this.channelId} Length:{this.length} Data:...";
    }
    public (nuint, bool) Channel()
    {
        return (this.channelId, true);
    }

    public byte[] MarshalMux()
    {
        // 8 + 1 + this.Data.Length
        var bufferSize = (int)PayloadSizes.MessageChannelData + 1 + this.length;
        var stream = new MemoryStream(new byte[bufferSize]);
        using (var writer = new BinaryWriter(stream, Encoding.BigEndianUnicode, false))
        {
            // 0
            writer.Write((byte)MessageType.MessageChannelData);
            // 1-5
            writer.Write(this.channelId);
            // 5-9
            writer.Write(this.length);
            // 10..
            writer.Write(this.data);
        }
        return stream.ToArray();
    }
    public void UnmarshalMux(byte[] b)
    {
        var stream = new MemoryStream(b);
        using (var reader = new BinaryReader(stream, Encoding.BigEndianUnicode))
        {
            // 0
            reader.ReadByte();
            // 1-5
            this.channelId = reader.ReadUInt32();
            // 5-9
            this.length = reader.ReadUInt32();
            // 10..
            this.data = reader.ReadBytes((int)this.length);
        }
    }
}